crates/holochain_zome_types/src/query.rs

```diff
@@ -1,26 +1,76 @@
 //! Types for source chain queries
 
+use std::collections::HashMap;
+use std::collections::HashSet;
+
 use crate::header::EntryType;
-use crate::header::Header;
 use crate::header::HeaderType;
 use crate::warrant::Warrant;
+use crate::Element;
+use crate::HeaderHashed;
+use holo_hash::EntryHash;
+use holo_hash::HasHash;
 use holo_hash::HeaderHash;
 pub use holochain_serialized_bytes::prelude::*;
 
+/// Defines several ways that queries can be restricted to a range.
+/// Notably hash bounded ranges disambiguate forks whereas sequence indexes do
+/// not as the same position can be found in many forks.
+/// The reason that this does NOT use native rust range traits is that the hash
+/// bounded queries MUST be inclusive otherwise the integrity and fork
+/// disambiguation logic is impossible. An exclusive range bound that does not
+/// include the final header tells us nothing about which fork to select
+/// between N forks of equal length that proceed it. With an inclusive hash
+/// bounded range the final header always points unambiguously at the "correct"
+/// fork that the range is over. Start hashes are not needed to provide this
+/// property so ranges can be hash terminted with a length of proceeding
+/// elements to return only. Technically the seq bounded ranges do not imply
+/// any fork disambiguation and so could be a range but for simplicity we left
+/// the API symmetrical in boundedness across all enum variants.
+/// @TODO It may be possible to provide/implement RangeBounds in the case that
+/// a full sequence of elements/headers is provided but it would need to be
+/// handled as inclusive first, to enforce the integrity of the query, then the
+/// exclusiveness achieved by simply removing the final element after the fact.
+#[derive(serde::Serialize, serde::Deserialize, PartialEq, Clone, Debug)]
+pub enum ChainQueryFilterSequenceRange {
```

Comment:
This shouldn't really be called a "sequence range" anymore since it can be a sequence range or a hash range.
```suggestion
pub enum ChainQueryFilterRange {
```